listener: add CloseOnExec option

muttleyxd opened the pull request:
This solves following situation:
- we have a server using a Pistache, using port 9080
  - one of endpoints uses fork() or some similar stuff to create a child process which runs in background (and can outlive its parent)
- endpoint gets called, child process is created
- server gets killed
- child lives on, blocking port 9080
- server cannot be restarted, because port 9080 is in use

example:
```
#include <cstdlib>

#include "pistache/endpoint.h"

using namespace Pistache;

class HelloHandler : public Http::Handler {
public:
    HTTP_PROTOTYPE(HelloHandler)

    void onRequest(const Http::Request&, Http::ResponseWriter response) override
    {
        std::system("sleep 1000 <&- &"); // fork magic - fire and forget style
        response.send(Http::Code::Ok, "success");
    }
};

int main()
{
    Pistache::Address addr(Pistache::Ipv4::any(), Pistache::Port(9080));
    auto opts = Pistache::Http::Endpoint::options().threads(1);
    // opts.flags(Pistache::Tcp::Options::CloseOnExec); // we can uncomment this to fix this behavior

    Http::Endpoint server(addr);
    server.init(opts);
    server.setHandler(Http::make_handler<HelloHandler>());
    server.serve();
}
```

```
$ cmake ..
...
$ make
...
$ ./cloexec
```

other terminal
```
$ curl 127.0.0.1:9080
success
$ sudo netstat -lepunt | grep 9080 
tcp        0      0 0.0.0.0:9080            0.0.0.0:*               LISTEN      1000       711080     98094/./cloexec
$ pkill -9 cloxec
$ sudo netstat -lepunt | grep 9080 #this would return nothing if we used CloseOnExec
tcp        0      0 0.0.0.0:9080            0.0.0.0:*               LISTEN      1000       711080     98359/sleep
```

kiplingw commented:
Thanks @muttleyxd. A maintainer will more carefully review your PR as soon as one of us has time. In the mean time, please ensure it's undergone all CI steps listed in our readme (including autopkgtesting).

muttleyxd commented:
Finally found some time for this, nested VMs can be a bit painful, but I managed to create one.
`Your branch and 'origin/add-cloexec' have diverged` is caused by rebasing my branch on current master.

![Test result](https://i.imgur.com/kOq406V.png)

Regarding autopkgtest:
In my opinion this should be done by CI of some kind. I know that VMs are not available with most CI providers, but TBH it's quite complicated setup and requires either good debugging skills or having deep knowledge about deb packaging.

Question: does `debian/rules` describe whole testing building and testing process? Or is there anything more?
If that's all then I could create and configure Docker/Drone combo (Drone is a free for FOSS, public CI/CD service) and avoid this whole VM setup for everyone wanting to change a simple thing. I know current VM setup is convenient for deb-based users, but anyone else has a hard time with this (instructions were a bit imprecise, had to create unsigned package and ofc all of this nested VM fun).

Are you interested in me making a PR with CI implementation for this? I guess it possible to do on Travis, though I don't know what are it's limitations regarding pipeline count/build times. If your Travis plan is full, then I'd go with Drone as it's perfect for this kind of thing.

kiplingw commented:
> Finally found some time for this, nested VMs can be a bit painful, but I managed to create one.
> `Your branch and 'origin/add-cloexec' have diverged` is caused by rebasing my branch on current master.

Just a tip, but you should avoid running anything as root unless you have to.

> Regarding autopkgtest:
> In my opinion this should be done by CI of some kind. I know that VMs are not available with most CI providers, but TBH it's quite complicated setup and requires either good debugging skills or having deep knowledge about deb packaging.

I partially agree with you. Autopkgtest isn't difficult to setup. It's only a handful of commands, all of which are laid out in our documentation. But yes, I do agree this should be done by CI.

> Question: does `debian/rules` describe whole testing building and testing process? Or is there anything more?

Actually `debian/rules` doesn't have anything directly related to CI. The whole standard is laid out in DEP-8 (which is suppose to be distro agnostic). This is implemented on a couple hundred Debian based distros as `autopkgtest` that's used with over [7,000 tests](https://fedoraproject.org/wiki/Changes/InvokingTestsPackaged#Comparison_with_Debian.27s_autopkgtest).

> If that's all then I could create and configure Docker/Drone combo (Drone is a free for FOSS, public CI/CD service) and avoid this whole VM setup for everyone wanting to change a simple thing. I know current VM setup is convenient for deb-based users, but anyone else has a hard time with this (instructions were a bit imprecise, had to create unsigned package and ofc all of this nested VM fun).

I think the issue was `debuild(1)` was missing the -uc flag, which I've just added. This is for unsigned changes which don't require a maintainer's key pair present.

> Are you interested in me making a PR with CI implementation for this? I guess it possible to do on Travis, though I don't know what are it's limitations regarding pipeline count/build times. If your Travis plan is full, then I'd go with Drone as it's perfect for this kind of thing.

If you find a CI service that supports DEP-8, it needs to:

- Integrate into our GitHub repository via hooks;
- Support all of our supported architectures, `amd64`, `arm64`, `armhf`, `i386`, `ppc64el`, and `s390x`;
- Support all recent Debian and Ubuntu releases.

The closest I've managed to find is [this](https://travis.debian.net/), but I think it still has [this](https://github.com/lamby/travis.debian.net/issues/64) problem. I think that might be better than nothing in at least testing on `amd64`. Eventually he may find a way to expand to other architectures. So yeah, give it a shot and thank you!

kiplingw commented:
Someone who is brilliant with this stuff who probably will have some useful insight is @pabs3. He's my go to guy whenever I need help with CI.

pabs3 commented:
I'm not entirely sure I fully understand all of this ticket, but some comments and a few somewhat related links:

As far as I know the only services that support DEP-8 are solely for packages within distributions; Debian has one ([debci](https://ci.debian.net/) and Ubuntu has one, not sure about other distributions though. debci supports several Debian architectures, but not all of them. 

Instead of trying to use a specific port in tests, which could run into conflicts if that port is in use by other processes running on the same system as the tests (think a developer machine also running the released version of the program), using [cwrap](https://cwrap.org/)'s [socket_wrapper](https://cwrap.org/socket_wrapper.html) would let you use a simulated port instead of a real one.

There are some [services](https://wiki.debian.org/Hardware/Wanted#Available_hardware) like the GCC compile farm providing gratis/paid access to non-amd64 physical hardware, some of that may be suitable for CI usage, but a lot of it will mainly be useful for manual testing. Possibly the [openSUSE instance of the open build service](https://en.opensuse.org/openSUSE:Build_Service) or Launchpad/Ubuntu PPAs could also be used for this.

kiplingw commented:
@pabs3, wise as always. The cwrap project has been very helpful in the past for me, in particular their `nss_wrapper`. The `socket_wrapper` would probably be useful here. 

We currently already have a Launchpad PPA for both stable and unstable releases, but Launchpad doesn't have any CI support for DEP-8 or anything else. The Open Build Service also doesn't support DEP-8. 

I think the best option we might have for the time being is the one I [suggested](https://travis.debian.net/) to him. It doesn't support other non-amd64 architectures (yet), but it is better than nothing.

pabs3 commented:
I just had an evil thought, you could add fake packages that use
autopkgtest to run the test suite instead of building something and
then "build" them on a PPA or OBS, you would have to manually trigger
rebuilds somehow though.

-- 
bye,
pabs

https://bonedaddy.net/pabs3/

dennisjenkins75 commented:
As far as the actual c++/sockets code changes, they are fine.  I'd like if a unit test were added though (mostly to verify that a future change does not break this functionality).  The test should verify both behaviors (close on exec not set, exec runs, socket still bound; and close on exec set, child does not hold socket).  Please look in "tests/" and select a reasonable place to add an additional test.  Note that most of our tests bind TCP port 0, then use an API to get the actual port that was bound.  This is so that the tests won't interfere with each other, and one can run multiple tests concurrently on the same machine, or on a machine that might have one of the (previously hardcoded) ports already in use.

muttleyxd commented:
> I'd like if a unit test were added though (mostly to verify that a future change does not break this functionality). The test should verify both behaviors (close on exec not set, exec runs, socket still bound; and close on exec set, child does not hold socket). 

Rebased and added a unit test to test this behavior (had to do some fork() magic to get this working).

muttleyxd commented:
Another force push - using `Pistache::Tcp::Options` instead of `Pistache::Flags<Pistache::Tcp::Options>` for passing options around, shortens the code by a little

kiplingw commented:
All tests pass. Thanks @muttleyxd for your contributions. Also feel free to help us integrate travis.debian.net.
